### PR TITLE
Fix regression in Delayed._length

### PR DESCRIPTION
**Pull request by crusaderky**
- Fixes regression introduced by #7177
- Closes dask/distributed#4493
- Supersedes #7192

FYI @jakirkham

### CODE REVIEW

File: dask/delayed.py [hunk 1]
```diff
@@ -510,17 +518,6 @@ def _rebuild(self, dsk, name=None):
         key = replace_name_in_key(self.key, name) if name else self.key
         return Delayed(key, dsk, self._length)
 
-    def __getstate__(self):
-        return tuple(getattr(self, i) for i in self.__slots__)
-
-    def __setstate__(self, state):
-        for k, v in zip(self.__slots__, state):
-            setattr(self, k, v)
```
Comment (jrbourbeau) [inline dask/delayed.py:518]: Why drop these methods? Are they equivalent to the default implementation?

File: dask/delayed.py [hunk 2]
```diff
@@ -532,26 +529,18 @@ def __dir__(self):
...
         return DelayedAttr(self, attr)
 
-    def __setattr__(self, attr, val):
-        if attr in self.__slots__:
-            object.__setattr__(self, attr, val)
-        else:
-            raise TypeError("Delayed objects are immutable")
-
-    def __setitem__(self, index, val):
-        raise TypeError("Delayed objects are immutable")
```
Comment (jrbourbeau) [inline dask/delayed.py:553]: Since `DelayedLeaf` and `DelayedAttr` are related to our internal implementation, users might better understand

```python
TypeError: Delayed objects are immutable
```

than

```python
TypeError: 'DelayedLeaf' object does not support item assignment
```

File: dask/tests/test_base.py [hunk 3]
```diff
@@ -1148,6 +1148,23 @@ def test_persist_delayed_change_name(key, new_name, new_key):
...
+
+
+def test_persist_delayedattr():
+    class C:
+        x = 1
+
+    x = delayed(C).x
+    (xx,) = persist(x)
+    assert isinstance(xx, Delayed)
+    assert xx.compute() == 1
```
Comment (jrbourbeau) [inline dask/tests/test_base.py:1165]: Thank you for adding these tests

File: dask/utils.py [hunk 4]
```diff
@@ -1056,6 +1056,8 @@ def ensure_dict(d: Mapping[K, V], *, copy: bool = False) -> Dict[K, V]:
 class OperatorMethodMixin(object):
     """A mixin for dynamically implementing operators"""
 
+    __slots__ = ()
```
Comment (jrbourbeau) [inline dask/utils.py:1059]: Why this change?

**Review by jrbourbeau — COMMENTED** [review]
FWIW I ran the `distributed` test suite with the changes here and everything passed (xref https://github.com/jrbourbeau/distributed/runs/1863749698)

File: dask/delayed.py (hunk 2 above)
Comment (crusaderky) [reply dask/delayed.py:553]: fair enough, I'll reinstate it

File: dask/utils.py (hunk 4 above)
Comment (crusaderky) [reply dask/utils.py:1059]: So that Delayed will no longer have a ``__dict__``. A class will have a dict unless itself and all of its parent classes declare ``__slots__``. The change is inconsequential for other classes that derive from the same mixin that don't define ``__slots__``.

File: dask/delayed.py (hunk 1 above)
Comment (crusaderky) [reply dask/delayed.py:518]: Yes. I suspect they were a leftover from Python 2, which needed this sort of massaging to pickle classes with ``__slots__``.

**Comment by jakirkham** [comment]
Would it make sense to have a test covering the Distributed use case or is that already covered here?

**Comment by jrbourbeau** [comment]
Since `persist` here and `client.persist` in `distributed` use the `__dask_postpersist__` and `__dask_keys__` interface, which is where the regression in https://github.com/dask/dask/pull/7177 was introduced, I think the tests added here which test those interfaces for `DelayedLeaf` and `DelayedAttr` should encompass the `distributed` case too

**Comment by jakirkham** [comment]
Thank you both! 😄
